src/arkode/arkode_mristep.c

```diff
@@ -1750,84 +1769,86 @@ int mriStep_TakeStepMRIGARK(ARKodeMem ark_mem, sunrealtype* dsmPtr, int* nflagPt
-     of the inner (fast) RHS function */
-  if (!(ark_mem->fn_is_current))
+     function here since it does not need to check for FSAL or SA methods and
+     thus avoids potentially unnecessary evaluations of the inner (fast) RHS function */
+  /*   compute the explicit component */
+  if (step_mem->explicit_rhs)
   {
-    /* compute the explicit component */
-    if (step_mem->explicit_rhs)
+    if (!ark_mem->fn_is_current || !step_mem->fse_is_current)
     {
-      retval = step_mem->fse(t0, ark_mem->yn, step_mem->Fse[0],
+      retval = step_mem->fse(ark_mem->tn, ark_mem->yn, step_mem->Fse[0],
                              ark_mem->user_data);
       step_mem->nfse++;
+      step_mem->fse_is_current = SUNTRUE;
       if (retval) { return ARK_RHSFUNC_FAIL; }
+
+      /* Add external forcing, if appropriate */
+      if (step_mem->expforcing)
+      {
+        step_mem->cvals[0] = ONE;
+        step_mem->Xvecs[0] = step_mem->Fse[0];
+        nvec               = 1;
+        mriStep_ApplyForcing(step_mem, ark_mem->tn, ONE, &nvec);
+        N_VLinearCombination(nvec, step_mem->cvals, step_mem->Xvecs,
+                             step_mem->Fse[0]);
+      }
     }
+  }
 
-    /* compute the implicit component */
-    if (step_mem->implicit_rhs)
+  /*   compute the implicit component */
+  if (step_mem->implicit_rhs)
+  {
+    if (!ark_mem->fn_is_current || !step_mem->fsi_is_current)
     {
-      retval = step_mem->fsi(t0, ark_mem->yn, step_mem->Fsi[0],
+      retval = step_mem->fsi(ark_mem->tn, ark_mem->yn, step_mem->Fsi[0],
                              ark_mem->user_data);
       step_mem->nfsi++;
+      step_mem->fsi_is_current = SUNTRUE;
       if (retval) { return ARK_RHSFUNC_FAIL; }
-    }
 
-    /* Add external forcing to Fse[0] or Fsi[0], if appropriate */
-    if (step_mem->expforcing)
-    {
-      step_mem->cvals[0] = ONE;
-      step_mem->Xvecs[0] = step_mem->Fse[0];
-      nvec               = 1;
-      mriStep_ApplyForcing(step_mem, t0, ONE, &nvec);
-      N_VLinearCombination(nvec, step_mem->cvals, step_mem->Xvecs,
-                           step_mem->Fse[0]);
-    }
-    if (step_mem->impforcing)
-    {
-      step_mem->cvals[0] = ONE;
-      step_mem->Xvecs[0] = step_mem->Fsi[0];
-      nvec               = 1;
-      mriStep_ApplyForcing(step_mem, t0, ONE, &nvec);
-      N_VLinearCombination(nvec, step_mem->cvals, step_mem->Xvecs,
-                           step_mem->Fsi[0]);
+      /* Add external forcing, if appropriate */
+      if (step_mem->impforcing)
+      {
+        step_mem->cvals[0] = ONE;
+        step_mem->Xvecs[0] = step_mem->Fsi[0];
+        nvec               = 1;
+        mriStep_ApplyForcing(step_mem, ark_mem->tn, ONE, &nvec);
+        N_VLinearCombination(nvec, step_mem->cvals, step_mem->Xvecs,
+                             step_mem->Fsi[0]);
+      }
     }
-
-    ark_mem->fn_is_current = SUNTRUE;
   }
+  ark_mem->fn_is_current = SUNTRUE;
```

Comment:
I'm not sure I fully understand your concern here, but I am confident that this suggestion is not the way to proceed.

`fn` is only ever used with the Hermite interpolation module, when temporal root-finding is enabled, or by steppers that wish to leverage its data directly (e.g., ERKStep); notably, `fn` is never directly used by MRIStep.  Thus, depending on the choice of interpolation module and temporal rootfinding, it is possible that `fn` may be `NULL`, so we should **not** copy data into it directly.

To ensure that I can understand your concern, let me explain what I think you're saying.  Since the TakeStep routines do not call `FullRHS` directly in `ARK_FULLRHS_START` mode, then `ark_mem->fn` may not be current at the start of the step.  Thus, anything *inside* the step that assumes `fn` is up-to-date will be accurate?  Given that we already disallow users from calling `ARKodeGetDky` from inside a current step, then I suppose the only real concern would be if the Hermite interpolation module is used for higher-order implicit predictors?  Are there other contexts where you think that `fn` needs to be current from inside the step calculation?

Separately, if `fn` may not be current at the start of the step, are you concerned that it might not be updated correctly at the end of the step (i.e., before `ARKodeGetDky` and temporal root-finding are used)?

Reply:
Yes, both of the cases you describe (using a high order predictor and updating data at the end of a step) are a concern and I think the only problematic use cases.

Suppose `fn_is_current` is false at the start of `TakeStep` (e.g., no dense output occurred between the last step and this step, so `FullRHS` was not called with the `END` mode), then the RHS functions will be called in `TakeStep` and `fn_is_current` is set to true (which is needed so we don't unnecessarily re-evaluate the RHS functions when re-attepting a failed step). However, the vector `ark_mem->fn` is not updated with the new RHS values so it still has an old value (like from the start of the previous step). So, when computing the high order predictor in the step or checking if `FullRHS` needs to be called to update the Hermite interpolation (or do root finding), the call to `FullRHS` will be skipped and the old `fn` value used instead. 

You're right that my suggestion is not correct since `fn` may not be allocated. So we should check that someone asked for `fn` by updating `if (!ark_mem->fn_is_current)` to `if (!ark_mem->fn_is_current && ark_mem->fn)` or
`if (!ark_mem->fn_is_current && ark_mem->call_fullrhs)`.

Additionally, since `fs*_is_current` is only set to false when the forcing changes we'll need to move `ark_mem->fn_is_current = SUNTRUE` outside the updated `if` so the flag is updated even if `fn` is not needed since it determines if the RHS functions should be called here. Related to this is the comment [here](https://github.com/LLNL/sundials/pull/564#discussion_r1846670283) about duplicating the stiffly accurate checks so a copy is done instead of an evaluation if possible.